pylib/anki/_backend/genbackend.py

```diff
@@ -129,52 +110,58 @@ def get_input_assign(msg):
     return ", ".join(f"{f.name}={f.name}" for f in fields)
 
 
+# messages we don't want to unroll in codegen
+SKIP_UNROLL_INPUT = {
+    "TranslateString",
+    "SetPreferences",
+    "UpdateDeckConfigs",
+    "AnswerCard",
+    "ChangeNotetype",
+    "CompleteTag",
+}
+SKIP_UNROLL_OUTPUT = {"GetPreferences"}
+
+
 def render_method(service_idx, method_idx, method):
     input_name = method.input_type.name
+
     if (
         (input_name.endswith("Request") or len(method.input_type.fields) < 2)
         and not method.input_type.oneofs
         and not method.name in SKIP_UNROLL_INPUT
     ):
-        input_args = get_input_args(method.input_type)
-        input_assign = get_input_assign(method.input_type)
-        input_assign_outer = (
-            f"input = {fullname(method.input_type.full_name)}({input_assign})\n        "
-        )
+        input_params = get_input_args(method.input_type)
+        input_assign_full = f"input = {fullname(method.input_type.full_name)}({get_input_assign(method.input_type)})"
     else:
-        input_args = f"self, input: {fullname(method.input_type.full_name)}"
-        input_assign_outer = ""
+        input_params = f"self, input: {fullname(method.input_type.full_name)}"
+        input_assign_full = ""
+
     name = fix_snakecase(stringcase.snakecase(method.name))
+
     if (
         len(method.output_type.fields) == 1
         and method.output_type.fields[0].type != TYPE_ENUM
         and method.name not in SKIP_UNROLL_OUTPUT
     ):
         # unwrap single return arg
         f = method.output_type.fields[0]
-        single_field = f".{f.name}"
         return_type = python_type(f)
+        single_attribute = f".{f.name}"
     else:
-        single_field = ""
         return_type = fullname(method.output_type.full_name)
-
-    if method.name in SKIP_DECODE:
-        return_type = "bytes"
-
-    buf = f"""\
-    def {name}({input_args}) -> {return_type}:
-        {input_assign_outer}"""
-
-    if method.name in SKIP_DECODE:
-        buf += f"""return self._run_command({service_idx}, {method_idx}, input)
+        single_attribute = ""
+
+    return f"""\
+    def {name}_raw(self, input: bytes) -> bytes:
+        return self._run_command({service_idx}, {method_idx}, input)
+
+    def {name}({input_params}) -> {return_type}:
+        {input_assign_full}
+        raw_bytes = self.{name}_raw(input.SerializeToString())
```

Comment:
But wouldn't this just be code duplication?

Reply:
It is code duplication, though it's a single line in our code generation script. self.foo() requires a dictionary lookup in the self object, and function calls in Python are relatively expensive. For most of our backend methods we probably wouldn't notice a difference either way, but it may offer some minor savings for simple methods that do not lock the backend, such as some of the string handling routines. Premature optimization perhaps, but it is only a single line we're talking about here :-)